parquet: decoder should handle empty bit(0) arrays

sql/importer.TestRandomParquetExports [failed](https://tanzanite.cluster.engflow.com/invocations/default/1e04a3aa-13a4-4cca-b923-2eeaffc4fadd?testReportRun=2&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc3FsL2ltcG9ydGVyOmltcG9ydGVyX3Rlc3Q=) on release-24.3 @ [112484d48f818e6e8585fa944a150d06e5acbbfb](https://github.com/cockroachdb/cockroach/commits/112484d48f818e6e8585fa944a150d06e5acbbfb):


```
=== RUN   TestRandomParquetExports
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestRandomParquetExports1839873803
    test_log_scope.go:76: use -show-logs to present logs inline
    test_server_shim.go:152: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    exportparquet_test.go:252: exporting as parquet from random table with schema: 
         [[table3 CREATE TABLE public.table3 (
        	col3_0 INT8 NOT NULL,
        	col3_1 INT8 NOT NULL,
        	col3_2 BIT(0)[] NOT NULL,
        	col3_3 BPCHAR NOT NULL,
        	col3_4 DECIMAL NOT NULL,
        	col3_6 DECIMAL NOT NULL,
        	col3_7 INT8 NOT NULL AS (abs(col3_1)) VIRTUAL,
        	CONSTRAINT table3_pkey PRIMARY KEY (col3_6 DESC, col3_2 DESC, col3_3 ASC),
        	FAMILY fam_0_col3_0_col3_1 (col3_0, col3_1),
        	FAMILY fam_1_col3_2 (col3_2),
        	FAMILY fam_2_col3_4 (col3_4),
        	FAMILY fam_4_col3_6_col3_3 (col3_6, col3_3)
        )]]
    exportparquet_test.go:168: 
        	Error Trace:	pkg/sql/importer/exportparquet_test.go:168
        	            				pkg/sql/importer/exportparquet_test.go:134
        	            				pkg/sql/importer/exportparquet_test.go:121
        	            				pkg/sql/importer/exportparquet_test.go:262
        	Error:      	Not equal: 
        	            	expected: &tree.DBitArray{BitArray:bitarray.BitArray{words:[]uint64{}, lastBitsUsed:0x0}}
        	            	actual  : &tree.DBitArray{BitArray:bitarray.BitArray{words:[]uint64(nil), lastBitsUsed:0x0}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -2,4 +2,3 @@
        	            	  BitArray: (bitarray.BitArray) {
        	            	-  words: ([]uint64) {
        	            	-  },
        	            	+  words: ([]uint64) <nil>,
        	            	   lastBitsUsed: (uint8) 0
        	Test:       	TestRandomParquetExports
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestRandomParquetExports1839873803
--- FAIL: TestRandomParquetExports (15.81s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=2</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #137673 sql/importer: TestRandomParquetExports failed [C-test-failure O-robot T-sql-queries branch-release-24.2 release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomParquetExports.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-45732

--- a/pkg/sql/export/exportparquet_test.go
+++ b/pkg/sql/export/exportparquet_test.go
@@ -229,22 +229,6 @@ func TestRandomParquetExports(t *testing.T) {
 					require.NoError(t, err)
 
 					for _, col := range cols {
-						// Skip tables with BIT(0) columns (or arrays of BIT(0)).
-						// The testutils parquet decoder cannot decode these as
-						// into the expected empty bit array and decodes the as
-						// nil. Ignore these to avoid these flakes.
-						colTyp := col.Typ
-						if colTyp.Family() == types.ArrayFamily {
-							colTyp = colTyp.ArrayContents()
-						}
-						if colTyp.Family() == types.BitFamily && colTyp.Width() == 0 {
-							// We skip these because of the aforementioned
-							// decoder issue.
-							t.Logf("skipping table '%s' because it has BIT(0) column '%s' (type: %s)",
-								tableName, col.Name, colTyp.String())
-							return fmt.Errorf("table has BIT(0) column")
-						}
-
 						// TODO(#104278): don't call this function to check if a type is supported.
 						// We should explicitly use the ones supported by  util/parquet).
 						_, err := parquet.NewSchema([]string{"test"}, []*types.T{col.Typ})

--- a/pkg/util/parquet/testutils.go
+++ b/pkg/util/parquet/testutils.go
@@ -419,6 +419,12 @@ func ValidateDatum(t *testing.T, expected tree.Datum, actual tree.Datum) {
 		require.Equal(t, expected.(*tree.DEnum).LogicalRep, actual.(*tree.DEnum).LogicalRep)
 	case types.CollatedStringFamily:
 		require.Equal(t, expected.(*tree.DCollatedString).Contents, actual.(*tree.DCollatedString).Contents)
+	case types.BitFamily:
+		// Compare using string representation to avoid nil vs empty slice
+		// mismatch. BIT(0) produces a BitArray with an empty words slice from
+		// SQL, but the parquet decoder's bitarray.Parse("") returns a nil words
+		// slice. Both are semantically equivalent.
+		require.Equal(t, expected.String(), actual.String())
 	case types.OidFamily:
 		require.Equal(t, expected.(*tree.DOid).Oid, actual.(*tree.DOid).Oid)
 	default:

--- a/pkg/util/parquet/writer_test.go
+++ b/pkg/util/parquet/writer_test.go
@@ -325,6 +325,28 @@ func TestBasicDatums(t *testing.T) {
 			},
 		},
 		{
+			// Regression test for #137757. The SQL layer produces BIT(0) values
+			// with a non-nil empty words slice, while the parquet decoder's
+			// bitarray.Parse("") returns a nil words slice. Both are semantically
+			// equivalent. Use FromEncodingParts to construct the expected value
+			// with a non-nil empty words slice, matching what the SQL layer
+			// produces.
+			name: "bitarray_empty",
+			sch: &colSchema{
+				columnTypes: []*types.T{types.MakeBit(0), types.MakeBit(0)},
+				columnNames: []string{"a", "b"},
+			},
+			datums: func() ([][]tree.Datum, error) {
+				ba, err := bitarray.FromEncodingParts([]uint64{}, 0)
+				if err != nil {
+					return nil, err
+				}
+				return [][]tree.Datum{
+					{&tree.DBitArray{BitArray: ba}, tree.DNull},
+				}, nil
+			},
+		},
+		{
 			name: "bytes",
 			sch: &colSchema{
 				columnTypes: []*types.T{types.Bytes, types.Bytes},

